faunadb-scala/src/main/scala/faunadb/FaunaClient.scala

```diff
@@ -248,50 +240,89 @@ class FaunaClient private (connection: Connection) {
-      val error = QueryErrorResponse(statusCode, parsedErrors)
-      error
-    }.recoverWith {
-      case e: FaunaException => Future.failed(e)
-      case unavailable if response.status().code() == 503 => Future.failed(new UnavailableException("Service Unavailable: Unparseable response.", unavailable))
-      case unknown => Future.failed(new UnknownException(s"Unparseable service $unknown response.", unknown))
-    }.andThen {
-      case _ => response.release()
+
+      def getErrors(body: JsonNode): Future[Iterator[JsonNode]] = Option(body.get("errors")) match {
+        case Some(errors) => Future(errors.asInstanceOf[ArrayNode].iterator().asScala)
+        case None => Future.successful(Iterator.empty)
+      }
+
+      def parseErrors(errors: Iterator[JsonNode]): Future[IndexedSeq[QueryError]] = Future {
+        errors.map(json.treeToValue(_, classOf[QueryError])).toIndexedSeq
+      }
+
+      val result: Future[QueryErrorResponse] =
+        for {
+          body <- parseResponseBody(response)
+          errors <- getErrors(body)
+          queryErrors <- parseErrors(errors)
+        } yield QueryErrorResponse(statusCode, queryErrors)
+
+      result
+        .recoverWith {
+          case e: FaunaException => Future.failed(e)
+          case unavailable if response.status().code() == 503 => Future.failed(new UnavailableException("Service Unavailable: Unparseable response.", unavailable))
+          case unknown => Future.failed(new UnknownException(s"Unparseable service $unknown response.", unknown))
+        }.andThen {
+          case _ => response.release()
+        }
     }
 
-    def parseErrorAndFailWith(fun: QueryErrorResponse => FaunaException): Future[Nothing] = {
-      parseError().flatMap { errors =>
+    def parseErrorsAndFailWith(fun: QueryErrorResponse => FaunaException): Future[Nothing] = {
+      parseErrors().flatMap { errors =>
         val exception = fun(errors)
         Future.failed(exception)
       }
     }
 
     response.status().code() match {
-      case 400 => parseErrorAndFailWith(new BadRequestException(_))
-      case 401 => parseErrorAndFailWith(new UnauthorizedException(_))
-      case 403 => parseErrorAndFailWith(new PermissionDeniedException(_))
-      case 404 => parseErrorAndFailWith(new NotFoundException(_))
-      case 500 => parseErrorAndFailWith(new InternalException(_))
-      case 503 => parseErrorAndFailWith(new UnavailableException(_))
-      case _   => parseErrorAndFailWith(new UnknownException(_))
+      case 400 => parseErrorsAndFailWith(new BadRequestException(_))
+      case 401 => parseErrorsAndFailWith(new UnauthorizedException(_))
+      case 403 => parseErrorsAndFailWith(new PermissionDeniedException(_))
+      case 404 => parseErrorsAndFailWith(new NotFoundException(_))
+      case 500 => parseErrorsAndFailWith(new InternalException(_))
+      case 503 => parseErrorsAndFailWith(new UnavailableException(_))
```

Comment:
@marrony didn't we add a `502` here in the past? (load balancing stuff)

Reply:
Just in case, it was not present in master before!

https://github.com/fauna/faunadb-jvm/blob/d087a89b27875b662fbd5ca648bea3a3b626254f/faunadb-scala/src/main/scala/faunadb/FaunaClient.scala#L270

Reply:
Yeap.. I didn't find either... as we already know this is a "common error status", can we add a match for it?

Reply:
Yes! I'd say we create another ticket for that to keep things organized! :) We need to add it to the Java client as well and decide which exception to throw!

Reply:
I've created following tickets:

[[scala] Match 502 error responses to a concrete Exception](https://faunadb.atlassian.net/browse/DRV-148)

[[java] Match 502 error responses to a concrete Exception](https://faunadb.atlassian.net/browse/DRV-149)